src/main/kotlin/grpc/AuthenticationInterceptor.kt

```diff
@@ -0,0 +1,298 @@
+         * If token parsing fails (e.g., malformed user ID), the call is closed with [Status.UNAUTHENTICATED].
+         *
+         * @param ReqT The request type.
+         * @param RespT The response type.
+         * @param token The JWT access token to validate and parse.
+         * @param call The current gRPC server call.
+         * @param headers The metadata headers associated with the call.
+         * @param next The next handler in the interceptor chain.
+         * @return A [ServerCall.Listener] to proceed with the call, or a no-op listener if authentication fails.
+         */
+        private fun <ReqT : Any?, RespT : Any?> attachUserIdAndContinue(
+            token: String,
+            call: ServerCall<ReqT?, RespT?>?,
+            headers: Metadata?,
+            next: ServerCallHandler<ReqT?, RespT?>?,
+        ): ServerCall.Listener<ReqT?> =
+            @Suppress("TooGenericExceptionCaught")
+            try {
+                val userId = UUID.fromString(JWTUtils.getUserId(token))
+                val context = Context.current().withValue(GrpcContext.USER_ID_CONTEXT_KEY, userId)
+                // Run the rest of the call within the new context, handling a potentially null `next`
+                context.call { next?.startCall(call, headers) } ?: emptyListener()
+            } catch (e: Exception) {
+                logger.warn(e) { "Failed to extract user ID from a seemingly valid token." }
+                call?.close(Status.UNAUTHENTICATED.withDescription("Invalid token payload"), Metadata())
+                emptyListener()
+            }
+
+        /**
+         * Attempts to refresh the access token using the provided refresh token.
+         *
+         * This method performs the following:
+         * 1. Validates the refresh token. If invalid or missing, it rejects the call with [Status.UNAUTHENTICATED],
+         *    and sends a `Set-Cookie` header to delete the invalid refresh token on the client side.
+         * 2. If valid, it issues a new access token, attaches it as a `Set-Cookie` trailer,
+         *    and continues the call with the user ID attached to the gRPC [Context].
+         * 3. If any unexpected errors occur during refresh, it closes the call with [Status.INTERNAL].
+         *
+         * @param ReqT The request type.
+         * @param RespT The response type.
+         * @param refreshToken The JWT refresh token to validate and use for issuing a new access token.
+         * @param call The current gRPC server call.
+         * @param headers The metadata headers associated with the call.
+         * @param next The next handler in the interceptor chain.
+         * @return A [ServerCall.Listener] to proceed with the call, or a no-op listener if token refresh fails.
+         */
+        private fun <ReqT : Any?, RespT : Any?> handleTokenRefresh(
+            refreshToken: String?,
+            call: ServerCall<ReqT?, RespT?>?,
```

Comment:
`call` is non-nullable at this point.
```suggestion
            call: ServerCall<ReqT?, RespT?>,
```